src/main/java/com/marklogic/mapreduce/utilities/InternalUtilities.java

```diff
@@ -612,4 +606,32 @@ public static void sleep(long millis) throws InterruptedException {
             }
         }
     }
+
+    private static Credentials getOutputCredentials(Configuration conf) {
+        String user = conf.get(OUTPUT_USERNAME, "");
+        String password = conf.get(OUTPUT_PASSWORD, "");
+        String apiKey = conf.get(OUTPUT_API_KEY);
+        String tokenEndpoint = conf.get(OUTPUT_TOKEN_ENDPOINT);
+        String grantType = conf.get(OUTPUT_GRANT_TYPE);
+        if (apiKey != null) {
+            return new Credentials(apiKey.toCharArray(), tokenEndpoint,
+                grantType);
+        } else {
+            return new Credentials(user, password.toCharArray());
+        }
+    }
+
+    private static Credentials getInputCredentials(Configuration conf) {
+        String user = conf.get(INPUT_USERNAME, "");
+        String password = conf.get(INPUT_PASSWORD, "");
+        String apiKey = conf.get(INPUT_API_KEY);
+        String tokenEndpoint = conf.get(INPUT_TOKEN_ENDPOINT);
+        String grantType = conf.get(INPUT_GRANT_TYPE);
+        if (apiKey != null) {
+            return new Credentials(apiKey.toCharArray(), tokenEndpoint,
+                grantType);
+        } else {
+            return new Credentials(user, password.toCharArray());
+        }
+    }
```

Comment:
Might consider making it a bit more DRY, especially if we were to add additional types of credentials besides these three. Something like:

    
```
    private static Credentials getTokenCredentials(Configuration conf) {
        String apiKey = conf.get(OUTPUT_API_KEY);
        String tokenEndpoint = conf.get(OUTPUT_TOKEN_ENDPOINT);
        String grantType = conf.get(OUTPUT_GRANT_TYPE);
        Credentials credentials = null;
        if (apiKey != null) {
            credentials = new Credentials(apiKey.toCharArray(), tokenEndpoint,
                grantType);
        }
        return credentials;
    }
    
    private static Credentials getUsernamePasswordCredentials(Conf conf, String usernameKey, String passwordKey) {
        String user = conf.get(usernameKey, "");
        String password = conf.get(passwordKey, "");
        return new Credentials(user, password.toCharArray());
    }
    
    private static Credentials getOutputCredentials(Configuration conf) {
        Credentials credentials = getTokenCredentials(conf);
        if (credentials == null) {
            credentials = getUsernamePasswordCredentials(conf, OUTPUT_USERNAME, OUTPUT_PASSWORD);
        }
        return credentials;
    }

    private static Credentials getInputCredentials(Configuration conf) {
        Credentials credentials = getTokenCredentials(conf);
        if (credentials == null) {
            credentials = getUsernamePasswordCredentials(conf, INPUT_USERNAME, OUTPUT_PASSWPRD);
        }
        return credentials;
    }
```

Reply:
Thank you Mads. This part of code needs further refactoring to split Credentials into different child classes to support multiple auth types (http basic/digest, kerberos, mlcloud token, oauth). I think it will be done in the next minor release.

Thanks,
Vanessa